lib/src/main/kotlin/konig/kontext/Utils.kt

```diff
@@ -0,0 +1,34 @@
+package konig.kontext
+
+import com.google.protobuf.Message
+import io.grpc.CallOptions
+import io.grpc.Metadata
+import io.grpc.Metadata.Key
+import io.grpc.stub.AbstractStub
+import kotlin.coroutines.coroutineContext
+import kotlin.reflect.full.functions
+import kotlinx.coroutines.withContext
+import org.konigsoftware.konig.kontext.CoroutineKonigKontext
+
+internal val GRPC_HEADER = Key.of("konig-kontext-bin", Metadata.BINARY_BYTE_MARSHALLER)
+val CALL_OPTIONS_KEY: CallOptions.Key<Message> = CallOptions.Key.create("konig-kontext-call-option")
+
+suspend inline fun <reified ServiceT : AbstractStub<ServiceT>, RespT> ServiceT.kall(crossinline fn: suspend ServiceT.() -> RespT): RespT {
+    val konigKontextValue = coroutineContext[CoroutineKonigKontext.Key]?.context
+
+    val clientWithKonigKontextCallOption: ServiceT? = runCatching {
+        ServiceT::class.functions.find { it.name == "withOption" }
+            ?.call(this, CALL_OPTIONS_KEY, konigKontextValue) as? ServiceT
+    }.getOrNull()
+
+    return if (konigKontextValue != null && clientWithKonigKontextCallOption != null) clientWithKonigKontextCallOption.fn() else fn()
+}
```

Comment:
I actually think there might be a way to avoid having to use this function by using GrpcContext. This would prevent users of the library having to wrap all of their client service calls in this function which is quite a bit of boiler plate, and subject to be missed in some services. I'd still like to merge the current API since it is working and is a good starting point, although I will continue to do research and see if we can remove this in a follow up change

Reply:
Yes I was able to remove this function entirely! See https://github.com/konigsoftware/konig-kontext/pull/2 for the updated changes. I'd still like to merge this as is however so that we have the evolution of the library preserved